File: app/src/main/java/me/zipi/navitotesla/ui/favorite/FavoriteDialogFragment.java
package me.zipi.navitotesla.ui.favorite;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.RadioGroup;
import android.widget.Spinner;
import android.widget.TextView;

import java.util.Calendar;
import java.util.List;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.lifecycle.ViewModelProvider;
import lombok.Setter;
import me.zipi.navitotesla.AppExecutors;
import me.zipi.navitotesla.databinding.FavoriteDialogFragmentBinding;
import me.zipi.navitotesla.db.AppDatabase;
import me.zipi.navitotesla.db.PoiAddressEntity;
import me.zipi.navitotesla.model.Poi;
import me.zipi.navitotesla.service.poifinder.PoiFinderFactory;
import me.zipi.navitotesla.util.AnalysisUtil;

public class FavoriteDialogFragment extends DialogFragment implements
        AdapterView.OnItemSelectedListener, View.OnClickListener, RadioGroup.OnCheckedChangeListener {

    @Nullable
    PoiArrayAdapter poiArrayAdapter;
    private String dest;
    @Setter
    private Runnable onDismissListener;
    private FavoriteDialogViewModel favoriteDialogViewModel;
    @Nullable
    private FavoriteDialogFragmentBinding binding;

    public FavoriteDialogFragment() {
        super();
    }

    public FavoriteDialogFragment(String dest) {
        super();
        this.dest = dest;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //setCancelable(false);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        favoriteDialogViewModel =
                new ViewModelProvider(this).get(FavoriteDialogViewModel.class);
        binding = FavoriteDialogFragmentBinding.inflate(inflater, container, false);

        poiArrayAdapter = new PoiArrayAdapter(getContext(), android.R.layout.simple_spinner_item);
        poiArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        Spinner spinner = binding.addressSelector;
        spinner.setAdapter(poiArrayAdapter);
        spinner.setOnItemSelectedListener(this);


        binding.btnDestSearch.setOnClickListener(this);
        binding.btnFavoriteDismiss.setOnClickListener(this);
        binding.btnFavoriteSave.setOnClickListener(this);
        binding.radioGroup.setOnCheckedChangeListener(this);
        favoriteDialogViewModel.getPoiList().observe(getViewLifecycleOwner(), (v) -> updateSpinner());

        return binding.getRoot();
    }

    @Override
    public void onResume() {

        super.onResume();
        if (this.dest != null && binding != null) {
            binding.txtDest.setText(dest);
            searchDest();
        }
    }


    @Override
    public void onDestroy() {
        super.onDestroy();

        binding = null;
        poiArrayAdapter = null;
        dest = null;

    }

    @Override
    public void onClick(View v) {
        if (binding == null) {
            return;
        }
        if (v.getId() == binding.btnDestSearch.getId()) {
            searchDest();
        } else if (v.getId() == binding.btnFavoriteDismiss.getId()) {
            dismiss();
        } else if (v.getId() == binding.btnFavoriteSave.getId()) {
            saveFavorite();
        }
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        if (binding == null || favoriteDialogViewModel.getPoiList().getValue() == null) {
            return;
        }
        Poi poi = favoriteDialogViewModel.getPoiList().getValue().get(position);
        favoriteDialogViewModel.getSelectedPoi().postValue(poi);
        if (binding.radioRoadAddress.isChecked()) {
            binding.txtAddress.setText(poi.getRoadAddress());
        } else if (binding.radioAddress.isChecked()) {
            binding.txtAddress.setText(poi.getAddress());
        } else {
            binding.txtAddress.setText(poi.getGpsAddress());
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {

    }

    private void saveFavorite() {
        if (binding == null) {
            return;
        }
        PoiAddressEntity entity = PoiAddressEntity.builder()
                .registered(true)
                .poi(binding.txtDest.getText().toString())
                .address(binding.txtAddress.getText().toString())
                .created(Calendar.getInstance().getTime())
                .build();

        AppExecutors.execute(() -> {
            AppDatabase.getInstance(getContext()).poiAddressDao().insertPoi(entity);
            if (getActivity() != null) {
                getActivity().runOnUiThread(this::dismiss);
                if (onDismissListener != null) {
                    onDismissListener.run();
                }
            }

        });

    }

    private void searchDest() {
        AppExecutors.execute(() -> {
            if (binding == null) {
                return;
            }
            try {
                List<Poi> pois = PoiFinderFactory.getKakaoPoiFinder().listPoiAddress(binding.txtDest.getText().toString().trim());
                favoriteDialogViewModel.getPoiList().postValue(pois);
            } catch (Exception e) {
                AnalysisUtil.recordException(e);
            }
        });
    }


    private void updateSpinner() {
        if (poiArrayAdapter != null) {
            poiArrayAdapter.clear();
            poiArrayAdapter.addAll(favoriteDialogViewModel.getPoiList().getValue());
            poiArrayAdapter.notifyDataSetChanged();
        }
    }


    @Override
    public void onCheckedChanged(RadioGroup group, int checkedId) {
        if (favoriteDialogViewModel.getSelectedPoi().getValue() == null || binding == null) {
            return;
        }
        if (binding.radioRoadAddress.getId() == group.getCheckedRadioButtonId()) {
            binding.txtAddress.setText(favoriteDialogViewModel.getSelectedPoi().getValue().getRoadAddress());
        } else if (binding.radioAddress.getId() == group.getCheckedRadioButtonId()) {
            binding.txtAddress.setText(favoriteDialogViewModel.getSelectedPoi().getValue().getAddress());
        } else {
            binding.txtAddress.setText(favoriteDialogViewModel.getSelectedPoi().getValue().getGpsAddress());
        }
    }

    private static class PoiArrayAdapter extends ArrayAdapter<Poi> {
        public PoiArrayAdapter(Context context, @LayoutRes int resource) {
            super(context, resource);
        }

        @NonNull
        @Override
        public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {

            TextView view = (TextView) super.getView(position, convertView, parent);
            view.setSingleLine(false);
            Poi poi = getItem(position);
            String shortAddress = poi.getRoadAddress();
            String[] addressSplit = shortAddress.split(" ");
            if (addressSplit.length > 3) {
                shortAddress = addressSplit[0] + " " + addressSplit[1] + " " + addressSplit[2];
            }
            String text = poi.getPoiName() + "\n" + shortAddress;
            view.setText(text);
            view.setTextSize(12);

            return view;
        }

        @Override
        public View getDropDownView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
            TextView view = (TextView) super.getDropDownView(position, convertView, parent);
            view.setSingleLine(false);
            Poi poi = getItem(position);
            String text = poi.getPoiName() + "\n" + poi.getRoadAddress();
            view.setText(text);
            view.setTextSize(12);

            return view;
        }
    }
}

File: app/src/main/kotlin/me/zipi/navitotesla/ui/favorite/FavoriteDialogFragment.kt
package me.zipi.navitotesla.ui.favorite

import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.RadioGroup
import android.widget.TextView
import androidx.annotation.LayoutRes
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import me.zipi.navitotesla.R
import me.zipi.navitotesla.databinding.FavoriteDialogFragmentBinding
import me.zipi.navitotesla.db.AppDatabase
import me.zipi.navitotesla.db.PoiAddressEntity
import me.zipi.navitotesla.model.Poi
import me.zipi.navitotesla.service.poifinder.KakaoPoiFinder
import me.zipi.navitotesla.util.AnalysisUtil
import java.util.Date

class FavoriteDialogFragment :
    DialogFragment,
    AdapterView.OnItemSelectedListener,
    View.OnClickListener,
    RadioGroup.OnCheckedChangeListener {
    private lateinit var poiArrayAdapter: PoiArrayAdapter
    private var dest: String? = null
    private var prefillPoi: Poi? = null

    var onDismissListener: Runnable? = null
    private lateinit var favoriteDialogViewModel: FavoriteDialogViewModel
    private lateinit var binding: FavoriteDialogFragmentBinding

    constructor() : super()
    constructor(dest: String?) : super() {
        this.dest = dest
    }
    constructor(prefill: Poi) : super() {
        this.prefillPoi = prefill
        this.dest = prefill.poiName
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        favoriteDialogViewModel = ViewModelProvider(this)[FavoriteDialogViewModel::class.java]
        binding = FavoriteDialogFragmentBinding.inflate(inflater, container, false)
        poiArrayAdapter = PoiArrayAdapter(context, android.R.layout.simple_spinner_item)
        poiArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        val spinner = binding.addressSelector
        spinner.adapter = poiArrayAdapter
        spinner.onItemSelectedListener = this
        binding.btnDestSearch.setOnClickListener(this)
        binding.btnFavoriteDismiss.setOnClickListener(this)
        binding.btnFavoriteSave.setOnClickListener(this)
        binding.radioGroup.setOnCheckedChangeListener(this)
        favoriteDialogViewModel.poiList.observe(viewLifecycleOwner) { updateSpinner() }
        return binding.root
    }

    override fun onStart() {
        super.onStart()
        dialog?.window?.let { window ->
            val metrics = resources.displayMetrics
            val isLandscape =
                resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE
            val widthFraction = if (isLandscape) 0.7f else 0.92f
            val width = (metrics.widthPixels * widthFraction).toInt()
            window.setLayout(width, WindowManager.LayoutParams.WRAP_CONTENT)
            window.setGravity(Gravity.CENTER)
        }
    }

    override fun onResume() {
        super.onResume()
        val prefill = prefillPoi
        if (prefill != null) {
            // recent → + 진입 — DB 의 PoiAddressEntity 가 가진 도로명/지번/GPS 를 그대로 사용.
            // selectedPoi 를 미리 채워서 라디오 변경 시 onCheckedChanged 가 해당 값을 textfield 에 prefill.
            binding.txtDest.setText(prefill.poiName)
            binding.txtAddress.setText(prefill.getRoadAddress())
            favoriteDialogViewModel.selectedPoi.postValue(prefill)
        } else if (dest != null) {
            binding.txtDest.setText(dest)
            searchDest()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        dest = null
        prefillPoi = null
    }

    override fun onClick(v: View) {
        when (v.id) {
            binding.btnDestSearch.id -> {
                searchDest()
            }

            binding.btnFavoriteDismiss.id -> {
                dismiss()
            }

            binding.btnFavoriteSave.id -> {
                saveFavorite()
            }
        }
    }

    override fun onItemSelected(
        parent: AdapterView<*>?,
        view: View,
        position: Int,
        id: Long,
    ) {
        val poi = favoriteDialogViewModel.poiList.value?.get(position) ?: return
        favoriteDialogViewModel.selectedPoi.postValue(poi)
        val checkedId =
            when {
                binding.radioRoadAddress.isChecked -> binding.radioRoadAddress.id
                binding.radioAddress.isChecked -> binding.radioAddress.id
                else -> binding.radioGps.id
            }
        valueForRadio(poi, checkedId)?.let { binding.txtAddress.setText(it) }
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {}

    private fun saveFavorite() {
        val poiName = binding.txtDest.text.toString().trim()
        if (poiName.isEmpty()) {
            AnalysisUtil.makeToast(
                context = context,
                text = getString(R.string.emptyPoiName),
                level = AnalysisUtil.ToastLevel.WARN,
            )
            return
        }
        val address = binding.txtAddress.text.toString().trim()
        if (address.isEmpty() || address == "null,null") {
            AnalysisUtil.makeToast(
                context = context,
                text = getString(R.string.addressNotFound),
                level = AnalysisUtil.ToastLevel.WARN,
            )
            return
        }
        // 즐겨찾기는 사용자가 마지막으로 본 textfield 값 (도로명/지번/gps 어느 모드든) 을 그대로
        // roadAddress 컬럼에 저장하고, share 시 그 값을 그대로 사용한다 — 라디오/selected 분기 없음.
        val entity =
            PoiAddressEntity(
                poi = poiName,
                packageName = "",
                roadAddress = address,
                registered = true,
                created = Date(),
            )
        viewLifecycleOwner.lifecycleScope.launch {
            val dao = AppDatabase.getInstance().poiAddressDao()
            val existing = dao.findRegisteredByPoi(entity.poi)
            if (existing != null) {
                AnalysisUtil.makeToast(
                    context = context,
                    text = getString(R.string.duplicatedPoiName),
                    level = AnalysisUtil.ToastLevel.WARN,
                )
                return@launch
            }
            dao.insertPoi(entity)
            withContext(Dispatchers.Main) {
                dismiss()
            }
            onDismissListener?.run()
        }
    }

    private fun searchDest() {
        viewLifecycleOwner.lifecycleScope.launch {
            try {
                val pois =
                    KakaoPoiFinder().listPoiAddress(
                        binding.txtDest.text.toString().trim(),
                    )
                favoriteDialogViewModel.poiList.postValue(pois)
            } catch (e: Exception) {
                AnalysisUtil.recordException(e)
            }
        }
    }

    private fun updateSpinner() {
        poiArrayAdapter.clear()
        poiArrayAdapter.addAll(favoriteDialogViewModel.poiList.value ?: listOf())
        poiArrayAdapter.notifyDataSetChanged()
    }

    override fun onCheckedChanged(
        group: RadioGroup,
        checkedId: Int,
    ) {
        val poi = favoriteDialogViewModel.selectedPoi.value ?: return
        valueForRadio(poi, group.checkedRadioButtonId)?.let { binding.txtAddress.setText(it) }
    }

    private fun valueForRadio(
        poi: Poi,
        radioId: Int,
    ): String? {
        val raw =
            when (radioId) {
                binding.radioRoadAddress.id -> poi.getRoadAddress()
                binding.radioAddress.id -> poi.getAddress()
                binding.radioGps.id -> poi.getGpsAddress()
                else -> return null
            }
        if (raw.isEmpty() || raw == "null,null") return null
        return raw
    }

    class PoiArrayAdapter(
        context: Context?,
        @LayoutRes resource: Int,
    ) : ArrayAdapter<Poi?>(
            requireNotNull(context),
            resource,
        ) {
        override fun getView(
            position: Int,
            convertView: View?,
            parent: ViewGroup,
        ): View {
            val view = super.getView(position, convertView, parent) as TextView
            view.isSingleLine = false
            val poi = getItem(position)
            var shortAddress = poi!!.getRoadAddress()
            val addressSplit = shortAddress.split(" ")
            if (addressSplit.size > 3) {
                shortAddress = addressSplit[0] + " " + addressSplit[1] + " " + addressSplit[2]
            }
            val text: String = poi.poiName + "\n" + shortAddress
            view.text = text
            view.textSize = 12f
            return view
        }

        override fun getDropDownView(
            position: Int,
            convertView: View?,
            parent: ViewGroup,
        ): View {
            val view = super.getDropDownView(position, convertView, parent) as TextView
            view.isSingleLine = false
            val poi = getItem(position)
            val text: String = poi!!.poiName + "\n" + poi.getRoadAddress()
            view.text = text
            view.textSize = 12f
            return view
        }
    }
}
